src/engine/strat_engine/dmdevice.rs

```diff
@@ -0,0 +1,142 @@
+
+use std::fmt;
+use std::fmt::Display;
+
+use super::super::super::engine::{EngineResult, EngineError, ErrorEnum, PoolUuid};
+
+const FORMAT_VERSION: u16 = 1;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Layer {
+    Flex,
+    Thin,
+    ThinPool,
+}
+
+impl Display for Layer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Layer::Flex => write!(f, "flex"),
+            Layer::Thin => write!(f, "thin"),
+            Layer::ThinPool => write!(f, "thinpool"),
+        }
+    }
+}
+
+pub enum FlexRole {
+    MetadataVolume,
+    ThinData,
+    ThinMeta,
+}
+
+impl Display for FlexRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            FlexRole::MetadataVolume => write!(f, "mdv"),
+            FlexRole::ThinData => write!(f, "thindata"),
+            FlexRole::ThinMeta => write!(f, "thinmeta"),
+        }
+    }
+}
+
+pub enum ThinRole {
+    Filesystem,
+}
+
+impl Display for ThinRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ThinRole::Filesystem => write!(f, "fs"),
+        }
+    }
+}
+
+pub enum ThinPoolRole {
+    Pool,
+}
+
+impl Display for ThinPoolRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ThinPoolRole::Pool => write!(f, "pool"),
+        }
+    }
+}
+
+pub enum Role {
+    Flex(FlexRole),
+    Thin(ThinRole),
+    ThinPool(ThinPoolRole),
+}
+
+impl Display for Role {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Role::Flex(ref fr) => write!(f, "{}", fr),
+            Role::Thin(ref tr) => write!(f, "{}", tr),
+            Role::ThinPool(ref tr) => write!(f, "{}", tr),
+        }
+    }
+}
+
+/// Format a name for the flex layer.
+fn format_flex_name(pool_uuid: &PoolUuid, role: FlexRole) -> String {
+    return format!("stratis-{}-{}-flex-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
+
+/// Format a name for the thin layer.
+fn format_thin_name(pool_uuid: &PoolUuid, role: ThinRole) -> String {
+    return format!("stratis-{}-{}-thin-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
+
+/// Format a name for the thin pool layer.
+fn format_thinpool_name(pool_uuid: &PoolUuid, role: ThinPoolRole) -> String {
+    return format!("stratis-{}-{}-thinpool-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
+
+/// Format a name for the given pool UUID, layer, and role.
+/// Return an error if argument values are incompatible.
+/// It might, for example, be possible to pass a role value that is not
+/// appropriate for the givne layer.
+pub fn format_name(pool_uuid: &PoolUuid, layer: Layer, role: Role) -> EngineResult<String> {
+    match layer {
```

Comment:
If you were ever using it dynamically, with unknown values, it could certainly fail. I've decided to assume, for now, that that will never happen, in my new commit.